### PR TITLE
Reproduce and fix off-by-one error in reorg logic (#4119)

### PR DESCRIPTION
**Pull request by nuttycom**
This attempts to reproduce #4119 using a simple chain split.It currently seems to trigger a different issue, an assertion failure in `CheckBlockIndex` when restarting

### CODE REVIEW

**Comment by LarryRuane** [comment]
I can reproduce the problem too. It seems to me the test is invalid, because if I'm reading it correctly, first you start node 2 with `NO_SAPLING`, which doesn't activate Sapling until height 100, then mine on that node out to 14+100 = 114. 

Then the test restarts the node but with `HAS_SAPLING` which has Sapling activating at height 15. I don't understand completely, but I think the node's understandably getting confused because its blockchain on disk has blocks 15-99 as Sprout, but now it's starting up expecting them to be Sapling.

**Comment by steven-ecc** [comment]
@ebfull Can you review this one for Kris?

File: src/main.cpp [hunk 1]
```diff
@@ -4748,20 +4748,23 @@ bool RewindBlockIndex(const CChainParams& chainparams, bool& clearWitnessCaches)
             *pindex->nCachedBranchId == CurrentEpochBranchId(pindex->nHeight, consensus);
     };
 
-    int nHeight = 1;
```
Comment (nuttycom) [inline src/main.cpp:4751]: This change is what requires most careful review, please.

File: qa/rpc-tests/sapling_rewind_check.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+        logging.info("Checking overwinter block propagation.")
+        assert_equal(self.nodes[0].getbestblockhash(), block14)
+        assert_equal(self.nodes[1].getbestblockhash(), block14)
+        assert_equal(self.nodes[2].getbestblockhash(), block14)
+        logging.info("All nodes are on overwinter.")
+
+        # Generate a network split longer than the maximum rewind length (99)
+        logging.info("Generating network split...")
+        self.is_network_split=True # split the network
```
Comment (daira) [inline qa/rpc-tests/sapling_rewind_check.py:69]:
```suggestion
        self.split_network()
```

File: qa/rpc-tests/sapling_rewind_check.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        logging.info("All nodes are on overwinter.")
+
+        # Generate a network split longer than the maximum rewind length (99)
+        logging.info("Generating network split...")
+        self.is_network_split=True # split the network 
+        self.nodes[0].generate(50) # generate into sapling
+        expected = self.nodes[0].getbestblockhash()
+
+        # generate blocks into sapling; if this is set to 60, the test passes.
+        self.nodes[2].generate(120)
```
Comment (daira) [inline qa/rpc-tests/sapling_rewind_check.py:74]: I don't understand why it's 120.

File: qa/rpc-tests/test_framework/mininode.py [hunk 4]
```diff
@@ -31,6 +31,10 @@
 import logging
 import copy
 from pyblake2 import blake2b
+from test_framework.util import \
+    SPROUT_BRANCH_ID, OVERWINTER_BRANCH_ID, \
+    SAPLING_BRANCH_ID, BLOSSOM_BRANCH_ID, HEARTWOOD_BRANCH_ID, \
+    NU4_BRANCH_ID
```
Comment (daira) [inline qa/rpc-tests/test_framework/mininode.py:37]:
```suggestion
from test_framework.util import (SPROUT_BRANCH_ID, OVERWINTER_BRANCH_ID,
    SAPLING_BRANCH_ID, BLOSSOM_BRANCH_ID, HEARTWOOD_BRANCH_ID,
    NU4_BRANCH_ID)
```

File: qa/rpc-tests/sapling_rewind_check.py (hunk 3 above)
Comment (nuttycom) [reply qa/rpc-tests/sapling_rewind_check.py:74]: It just needs to be anything > 90. It would probably be worthwhile to parameterize this test and make sure that the boundary conditions around the cutoff are all correct.

**Comment by nuttycom** [comment]
@zkbot r+

**Comment by nuttycom** [comment]
@zkbot r+

**Comment by nuttycom** [comment]
@zkbot try

**Comment by str4d** [comment]
Failed due to pyflakes errors.

**Comment by str4d** [comment]
Ah, I didn't spot the fix is already there.

@zkbot r+
